src/cls/rgw/cls_rgw.cc

```diff
@@ -2501,12 +2495,52 @@ static int list_plain_entries(cls_method_context_t hctx,
     if (count >= (int)max) {
       return count;
     }
-    start_after_key = entry.idx;
   }
 
   return count;
 }
 
+static int list_plain_entries(cls_method_context_t hctx,
+                              const string& name,
+                              const string& marker,
+                              uint32_t max,
+                              list<rgw_cls_bi_entry> *entries,
+                              bool *pmore) {
+  string start_after_key = marker;
+  string end_key;
+  bi_log_prefix(end_key);
+  int r;
+  bool more;
+
+  if (start_after_key < end_key) {
+    // listing ascii plain namespace
+    int r = list_plain_entries(hctx, name, start_after_key, end_key, max,
+                               entries, &more);
+    if (r < 0) {
+      return r;
+    }
+    if (r >= (int)max || !more) {
+      if (pmore) {
+	*pmore = more;
+      }
+      return r;
+    }
+    start_after_key = BI_PREFIX_END;
+    max = max - r;
```

Comment:
Given that you reduce max whenever you call list_plain_entries twice, I'm unclear how we're returning twice as many entries back.

Reply:
This `max` (provided by the caller) has nothing to do with `osd_max_omap_entries_per_request` config option, which `cls_cxx_map_get_vals` respects. I.e. if `osd_max_omap_entries_per_request` is less than `max`, `cls_cxx_map_get_vals` will limit the list to this value.

And `cls_rgw.bi_lis` tests this. It may set `max` higher than `osd_max_omap_entries_per_request`, and expects that only ``osd_max_omap_entries_per_request` entries will be returned. But now, due to `cls_cxx_map_get_vals` may be called twice, `bi_lis` may return more.

Reply:
Thanks for the explanation. In that case I think the test is wrong. It shouldn't be testing that low-level parameter with a high-level call.